### PR TITLE
PMM-6347 Handle kubeconfig.

### PR DESCRIPTION
**Pull request by askomorokhov**
https://jira.percona.com/browse/PMM-6347

```bash
docker run \
    --rm \
    --mount='type=bind,src=/Users/als/go/src/github.com/percona-platform/dbaas-api,dst=/work' \
    --net=host docker.pkg.github.com/percona-platform/dbaas-api/dbaas-prototool:latest \ 
    prototool grpc \
    --address host.docker.internal:20201  \
    --method percona.platform.dbaas.controller.v1beta1.XtraDBClusterAPI/ListXtraDBClusters \
    --data "{\"kube_auth\": {\"kubeconfig\": `kubectl config view -o json | jq -c | jq -R`}}"
```

FB: https://github.com/Percona-Lab/pmm-submodules/pull/996

### CODE REVIEW

File: service/cluster/xtra_db_cluster.go [hunk 1]
```diff
@@ -41,10 +41,10 @@ func New(p *message.Printer) *Service {
 
 // ListXtraDBClusters returns a list of XtraDB clusters.
 func (s *Service) ListXtraDBClusters(ctx context.Context, req *controllerv1beta1.ListXtraDBClustersRequest) (*controllerv1beta1.ListXtraDBClustersResponse, error) {
-	client := k8sclient.NewK8Client(logger.Get(ctx))
+	client := k8sclient.NewK8Client(logger.Get(ctx), req.KubeAuth.Kubeconfig)
 	defer client.Cleanup()
-
-	return nil, status.Error(codes.Unimplemented, s.p.Sprintf("not implemented"))
+	xtraDBClusters, err := client.ListXtraDBClusters(ctx)
```
Comment (AlekSi) [inline service/cluster/xtra_db_cluster.go:46]: @BupycHuk is working on this right now; be ready for merge conflicts

File: service/k8sclient/kubectl/kubectl.go [hunk 2]
```diff
@@ -34,33 +38,54 @@ import (
...
-	path, err := exec.LookPath(cmd[0])
-	l.Debugf("path: %s, err: %v", path, err)
+	kubectlPath, err := exec.LookPath(cmd[0])
+	l.Debugf("kubectlPath: %s, err: %v", kubectlPath, err)
 	if e, ok := err.(*exec.Error); err != nil && ok && e.Err == exec.ErrNotFound {
 		cmd = []string{"minikube", "kubectl", "--"}
 	}
+
+	// Handle kubeconfig.
+	tmpDir, err := ioutil.TempDir("/tmp", "dbaas-controller-kubeconfigs-")
```
Comment (AlekSi) [inline service/k8sclient/kubectl/kubectl.go:60]: I think the first argument can be empty

File: service/k8sclient/kubectl/kubectl.go [hunk 3]
```diff
@@ -34,33 +38,54 @@ import (
...
+	if err != nil {
+		l.Fatal(err)
+	}
+
+	kubeconfigPath := path.Join(tmpDir, kubeconfigFileName)
+
+	err = ioutil.WriteFile(kubeconfigPath, []byte(kubeconfig), 0600)
```
Comment (AlekSi) [inline service/k8sclient/kubectl/kubectl.go:67]: Go 1.14 added a new syntax for octal numbers: `0o600`. Let use it

File: service/k8sclient/kubectl/kubectl.go [hunk 4]
```diff
@@ -34,33 +38,54 @@ import (
...
+	kubectlPath, err := exec.LookPath(cmd[0])
+	l.Debugf("kubectlPath: %s, err: %v", kubectlPath, err)
 	if e, ok := err.(*exec.Error); err != nil && ok && e.Err == exec.ErrNotFound {
 		cmd = []string{"minikube", "kubectl", "--"}
 	}
+
+	// Handle kubeconfig.
+	tmpDir, err := ioutil.TempDir("/tmp", "dbaas-controller-kubeconfigs-")
+	if err != nil {
+		l.Fatal(err)
```
Comment (AlekSi) [inline service/k8sclient/kubectl/kubectl.go:62]: This function should return error, not crash the process

File: service/k8sclient/kubectl/kubectl.go [hunk 5]
```diff
@@ -34,33 +38,54 @@ import (
...
 	}
 }
 
 // Cleanup removes temporary files created by that object.
 func (k *KubeCtl) Cleanup() {
-	// TODO Remove kubeconfig file https://jira.percona.com/browse/PMM-6347
+	err := os.RemoveAll(k.tmpDir)
+	if err != nil {
+		k.l.Errorf("cannot cleanup:%s", err)
+	}
```
Comment (AlekSi) [inline service/k8sclient/kubectl/kubectl.go:87]: `return`

File: service/k8sclient/kubectl/kubectl.go [hunk 6]
```diff
@@ -34,33 +38,54 @@ import (
...
+		tmpDir: tmpDir,
 	}
 }
 
 // Cleanup removes temporary files created by that object.
 func (k *KubeCtl) Cleanup() {
-	// TODO Remove kubeconfig file https://jira.percona.com/browse/PMM-6347
+	err := os.RemoveAll(k.tmpDir)
+	if err != nil {
+		k.l.Errorf("cannot cleanup:%s", err)
```
Comment (AlekSi) [inline service/k8sclient/kubectl/kubectl.go:86]: `"Cannot cleanup: %s"`

File: service/k8sclient/kubectl/kubectl_test.go [hunk 7]
```diff
@@ -0,0 +1,86 @@
...
+	"github.com/stretchr/testify/assert"
+	"github.com/stretchr/testify/require"
+
+	"github.com/percona-platform/dbaas-controller/utils/app"
+	"github.com/percona-platform/dbaas-controller/utils/logger"
+)
+
+func SetUp(t *testing.T) string {
+	cmd := []string{"dbaas-kubectl-1.16"}
+	kubectlPath, err := exec.LookPath(cmd[0])
```
Comment (AlekSi) [inline service/k8sclient/kubectl/kubectl_test.go:37]: Let's finish https://github.com/percona-platform/dbaas-controller/pull/34 first then

File: service/k8sclient/kubectl/kubectl_test.go [hunk 8]
```diff
@@ -19,12 +19,60 @@ package kubectl
...
+			}
+		}
+
+		assert.True(t, strings.HasSuffix(kubeconfigFlag, kubeconfigFileName))
+
+		kubeconfigFilePath := strings.Split(kubeconfigFlag, "=")[1]
+		dat, err := ioutil.ReadFile(kubeconfigFilePath) //nolint:gosec
+		require.NoError(t, err)
+		sha256KubeconfigActual := sha256.Sum256(dat)
+		assert.Equal(t, sha256KubeconfigExpected, sha256KubeconfigActual)
```
Comment (BupycHuk) [inline service/k8sclient/kubectl/kubectl_test.go:62]: why do we compare hash sum instead of body of config?

Comment (askomorokhov) [reply service/k8sclient/kubectl/kubectl_test.go:62]: Had the intention to reuse it in other test cases. But, here we have only one. Agree, following YAGNI - I'll remove it.

File: service/k8sclient/kubectl/kubectl.go [hunk 9]
```diff
@@ -58,14 +64,46 @@ func NewKubeCtl(ctx context.Context) (*KubeCtl, error) {
 
 	l.Infof("Using %q", strings.Join(cmd, " "))
 
-	// TODO Handle kubeconfig https://jira.percona.com/browse/PMM-6347
+	if kubeconfig == "" {
```
[the remark this answers is not in the document]
Comment (askomorokhov) [reply service/k8sclient/kubectl/kubectl.go:67]: We use this function outside API calls too, and additional validation covers these cases.
The empty kubeconfig is more an "exception" from the main flow that's why I put it in the "indented" flow.

[the remark this answers is not in the document]
Comment (BupycHuk) [reply service/k8sclient/kubectl/kubectl.go:67]: then ok

**Pull request merged**